cpp/ql/lib/semmle/code/cpp/models/implementations/GetText.qll

```diff
@@ -0,0 +1,35 @@
+import semmle.code.cpp.models.interfaces.DataFlow
+
+/**
+ * Returns the transated text index for a given gettext function `f`
+ */
+private int getTextArg(Function f) {
+  // basic variations of gettext
+  f.hasGlobalOrStdName("gettext") and result = 0
+  or
+  f.hasGlobalOrStdName("dgettext") and result = 1
+  or
+  f.hasGlobalOrStdName("dcgettext") and result = 1
+  or
+  // plural variations of gettext that take one format string for singular and another for plural form
+  f.hasGlobalOrStdName("ngettext") and
+  (result = 0 or result = 1)
+  or
+  f.hasGlobalOrStdName("dngettext") and
+  (result = 1 or result = 2)
+  or
+  f.hasGlobalOrStdName("dcngettext") and
+  (result = 1 or result = 2)
+}
+
+class GetTextFunction extends DataFlowFunction {
+  int argInd;
+
+  GetTextFunction() { argInd = getTextArg(this) }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(argInd) and output.isReturnValue()
```

Comment:
This should _not_ have argument-to-return-value flow. It should have argument-to-return-value flow. Since the pointer returned by `gettext` is completely unrelated to the pointer passed into the function it doesn't make any sense for this function to have argument-to-return-value flow.

Reply:
You mean it should have indirect argument to return value flow?

Reply:
Ben has already specified that the model has indirect argument to indirect return flow with this line:
> input.isParameterDeref(argInd) and output.isReturnValueDeref()

So nothing else is needed apart from that line.

Reply:
Fixed

Reply:
QQ: Does removing the line have any side effect if the input parameter were a constant literal? This kind of scenario has resulted in false negatives in the past where I have had to differentiate asExpr vs asIndirectExpr, and I'm not sure if that prior issue/concern is relevant here.

Reply:
That shouldn't have any effect, no.